### DISCUSSION TITLE
How validating data works?

### DISCUSSION THREAD

**Discussion by cpard** [cue-lang/cue#1824] [Q&A]
Hey everyone, 

This is the first time I'm trying to write something serious with CUE so I might be doing something terribly wrong here, but there's a behavior (that I will describe with examples) that I can't understand and it would be wonderful to get some feedback from the much more experienced people here.

Consider the following:

```
#lineitem: {
    returnflag: string | *"l_returnflag"
    linestatus: string | *"l_linestatus"
}

#column: {
    name: string
    value: bool | string | number
}

#result: [...#column]

lineitem: #lineitem & {
    returnflag: "return_flag"
    
    }

#q1_validation: #result & [
    {
        name: lineitem.linestatus
        value: "F"
    },
        {
        name: lineitem.returnflag
        value: "A"
    },
]

q1_test: #q1_validation & [
    {
        name: "this is wrong"
        value: "F"
    },
    {
        name: "return_flag"
        value: "A"
    }
]
```

when I cue eval the above, everything works fine, while I would expect that I would get a validation error because "this is wrong" != "l_linestatus"

The above behavior changes though if I add a definition for linestatus in lineitem. My assumption here is that as lineitem is of type #lineitem, it would also "inherit" the default values from it, which in a way it does in terms of what cue eval prints but it does not in terms of validating the value of the field.

Why is this happening?

Thanks in advance!!

**Comment by jkasky** [comment]
Hi @cpard,

I'm also in process of learning CUE but I think I may see something in your example you provided.

```CUE
// on line 3, this means linestatus can be any string and if not present defaults to "l_linestatus" 
linestatus: string | *"l_linestatus"

// [1] if linestatus should always be the value defined in #lineitem then you could change it to this
linestatus: "l_linestatus"

// [2] or you could place a constraint on #q1_validation using unification like this
#q1_validation: #result & [
    {
        name: lineitem.linestatus & "l_linestatus_for_q1_only"
        value: "F"
    },
        {
        name: lineitem.returnflag
        value: "A"
    },
]
```

If you run eval on [1] or [2] you'll see the validation fail at different points because the constraint is coming from different fields:

[1]
```shell
$ cue eval scratch.cue
q1_test.0.name: conflicting values "this is wrong" and "l_linestatus":
    ./scratch.cue:3:17
    ./scratch.cue:20:15
    ./scratch.cue:29:10
    ./scratch.cue:31:15
```

[2]
```shell
$ cue eval scratch.cue
q1_test.0.name: 2 errors in empty disjunction:
q1_test.0.name: conflicting values "l_linestatus" and "l_linestatus_for_q1_only":
    ./scratch.cue:3:27
    ./scratch.cue:20:15
    ./scratch.cue:20:37
    ./scratch.cue:29:10
q1_test.0.name: conflicting values "this is wrong" and "l_linestatus_for_q1_only":
    ./scratch.cue:20:37
    ./scratch.cue:29:10
    ./scratch.cue:31:15 
```

Now if you change q1_test.name from "this is wrong" to "l_linestatus_for_q1_only" and eval:
```shell
$ cue eval scratch.cue
#lineitem: {
    returnflag: "l_returnflag"
    linestatus: "l_linestatus"
}
#column: {
    name:  string
    value: bool | string | number
}
#result: []
lineitem: {
    returnflag: "return_flag"
    linestatus: "l_linestatus"
}
#q1_validation: [{
    name:  "l_linestatus_for_q1_only"
    value: "F"
}, {
    name:  "return_flag"
    value: "A"
}]
q1_test: [{
    name:  "l_linestatus_for_q1_only"
    value: "F"
}, {
    name:  "return_flag"
    value: "A"
}]
```

**Reply by cpard** [reply]
hey @jkasky thanks for your comments! btw, I kind of figured it out. If you make the following change:

#q1_validation: #result & [
    {
        name: "\(lineitem.linestatus)"
        value: "F"
    },
    {
        name: "\(lineitem.returnflag)"
        value: "A"
    }

In the validation type (btw is it correct to call it a type??) you will get the behavior I was expecting. But I still don't get why :). I'd love to understand why when I interpolate the value it works but when I refer to the field type/value from another definition directly, it doesn't when the definition is not explicit in the instance.

**Reply by jkasky** [reply]
I'm still learning so I wouldn't be able to say with 100% certainty that is correct to call it a type but from https://cuelang.org/docs/concepts/logic/#types-are-values to call it a type sounds correct but not the complete story as a type in CUE is also a value.

From my few days of experimenting and reading this is how I understand that to work.  When you use interpolation you are setting name in q1_validation to a string that is the result of the interpolation. So for the unification of q1_test & q1_validation - q1_validation is specifying that name must be equal to "\(lineitem.linestatus)". Without the interpolation setting name to `lineitem.linestatus` is a reference. And a reference is a copy of the expression from the field it is bound to. So the constraint in that case becomes the disjunction of `string | *"l_linestatus"` which allows any string value.

**Reply by cpard** [reply]
That makes sense, what makes it a bit confusing though is that when you eval you see the default value getting there instead of the disjunction so it's easy to assume that the default value got assigned and the constraint now becomes that value instead. But I think that what you are saying is also what is happening here!
